crates/admin/src/cluster_controller/service/state.rs

```diff
@@ -277,80 +281,779 @@ where
                 }
             }
         }
 
-        for (partition_id, persisted_lsns) in persisted_lsns_per_partition.into_iter() {
-            let log_id = LogId::from(partition_id);
-
-            // todo: Remove once Restate nodes can share partition processor snapshots
-            // only try to trim if we know about the persisted lsns of all known nodes; otherwise we
-            // risk that a node cannot fully replay the log; this assumes that no new nodes join the
-            // cluster after the first trimming has happened
-            if persisted_lsns.len() >= cluster_state.nodes.len() {
-                let min_persisted_lsn = persisted_lsns.into_values().min().unwrap_or(Lsn::INVALID);
-                // trim point is before the oldest record
-                let current_trim_point = self.bifrost.get_trim_point(log_id).await?;
-
-                if min_persisted_lsn >= current_trim_point + self.log_trim_threshold {
-                    debug!(
-                    "Automatic trim log '{log_id}' for all records before='{min_persisted_lsn}'"
-                );
-                    self.bifrost.admin().trim(log_id, min_persisted_lsn).await?
+        match snapshots_repository_configured
+            || Self::any_processors_report_archived_lsn(&cluster_state)
+        {
+            true => TrimMode::ArchivedLsn { partition_statuses },
+            false => TrimMode::PersistedLsn {
+                cluster_state,
+                partition_statuses,
+            },
+        }
+    }
+
+    /// Compute the safe trim points for each log, assuming that partitions are mapped to logs 1:1.
+    /// For a given cluster state and known log trim points, determines a new set of trim points for
+    /// those logs which are deemed safe to trim. Only logs that need action will contain an entry in
+    /// the result.
+    ///
+    /// The presence of any dead or suspect nodes in the cluster state struct will prevent logs from
+    /// being trimmed if archived LSN is *not* reported for all known partitions. Conversely, as long
+    /// as archived LSNs are reported for all partitions, trimming can continue even in the presence of
+    /// some dead nodes. This is because we assume that if those nodes are only temporarily down, they
+    /// can fast-forward state from the snapshot repository when they return into service.
+    fn safe_trim_points(
+        &self,
+        current_trim_points: &HashMap<LogId, Lsn>,
+        threshold: Option<Lsn>,
+    ) -> BTreeMap<LogId, (Lsn, PartitionId)> {
+        let mut safe_trim_points = BTreeMap::new();
+
+        if self.is_trimming_suspended() {
+            warn!(
+                "Log trimming is suspended until we can determine the processor state on all known cluster nodes. \
+                This may result in increased log usage. Prune permanently decommissioned nodes and/or enable partition \
+                snapshotting to unblock trimming."
+            );
```

Comment:
it would have been really nice if this log message tells us which nodes are dead/suspect.